### PR TITLE
[Structured Output][Refactor] Move `apply_grammar_bitmask()` method from `ModelRunner` to structured output utils

### PR DESCRIPTION
**Pull request by shen-shanshan**
# Essential Elements of an Effective PR Description Checklist

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.

## Purpose

Currently, the feature **structured output** is closely coupled with the `model_runner`, and we need to implement duplicate `apply_grammar_bitmask()` method in different `model_runner` of each platform, e.g., `gpu_model_runner`, `npu_model_runner`. Once there are changes have made in this method, we need to update the method in all kinds of `model_runner` to sync these changes.

Thus, maybe it's better to move these structured output related code in `model_runner` to the `structured_output` module to make it clearer and more extensible.

### CODE REVIEW

**Comment by shen-shanshan** [comment]
CC: @russellb

**Review by russellb — COMMENTED** [review]
The StructuredOutputManager isn't an object used by the model runner right now. How about just making it a utility function somewhere instead of hanging it off of that class? Maybe here https://github.com/vllm-project/vllm/blob/main/vllm/v1/structured_output/utils.py

**Comment by shen-shanshan** [comment]
> The StructuredOutputManager isn't an object used by the model runner right now. How about just making it a utility function somewhere instead of hanging it off of that class? Maybe here https://github.com/vllm-project/vllm/blob/main/vllm/v1/structured_output/utils.py

Thanks for your suggestion! I will modify it later~

**Comment by shen-shanshan** [comment]
@russellb I have updated it 😄 (though the CI get some errors that seems not relative to the changes in this PR).

**Comment by russellb** [comment]
I'm good with this in general, but it will need to be updated one more time.

It's in conflict with `main`. Before fixing it though, there are a few bug fix PRs that I'd like to merge first. Once they go in, if you can incorporate them into your change, I think we'll be good.

1. https://github.com/vllm-project/vllm/pull/22896
2. https://github.com/vllm-project/vllm/pull/22988
3. https://github.com/vllm-project/vllm/pull/22963

**Comment by shen-shanshan** [comment]
> I'm good with this in general, but it will need to be updated one more time.
> 
> It's in conflict with `main`. Before fixing it though, there are a few bug fix PRs that I'd like to merge first. Once they go in, if you can incorporate them into your change, I think we'll be good.
> 
> 1. [[FIXBUG] Correctly Apply Grammar Bitmask in Mixed Batches #22896](https://github.com/vllm-project/vllm/pull/22896)
> 2. [Upgrade xgrammar to 0.1.23 #22988](https://github.com/vllm-project/vllm/pull/22988)
> 3. [[Structured Outputs] [Bug] Fix misalignment in apply_grammar_bitmask causing unintended masking and NaN logits #22963](https://github.com/vllm-project/vllm/pull/22963)

OK, no problem. After these fix PR have been merged, I will update this soon.

**Comment by shen-shanshan** [comment]
@russellb Hello, all the fix PRs you mentioned have already been merged, and I have rebased on the latest code. 😃

**Review by russellb — APPROVED** [review]
lgtm, thanks!

**Comment by shen-shanshan** [comment]
@russellb The CI run failed due to https://github.com/vllm-project/vllm/pull/24366, and I have rebased once again. Now the CI has all passed.

**Review by benchislett — APPROVED** [review]
LGTM
